src/lib.rs

```diff
@@ -563,3 +584,145 @@ impl Encoder for URIReserved {
         }
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum IriSegmentKind {
+    Segment,
+    SegmentNoColons,
+    Fragment,
+    Query,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct IriReserved(IriSegmentKind);
+
+impl Encoder for IriReserved {
+    fn encode(&self, c: char) -> bool {
+        // iunreserved
+        if c.is_ascii_alphanumeric() {
+            return false;
+        }
+
+        match c {
+            // ipchar
+            '@' => return false,
+            // iunreserved
+            '-' | '.' | '_' | '~' => return false,
+            // sub-delims
+            '!' | '$' | '&' | '\'' | '(' | ')' | '*' | '+' | ',' | ';' | '=' => return false,
+            '/' | '?' => {
+                return self.0 != IriSegmentKind::Query && self.0 != IriSegmentKind::Fragment
+            }
+            ':' => return self.0 == IriSegmentKind::SegmentNoColons,
+            _ => { /* fall through */ }
+        }
+
+        match c as u32 {
+            // ucschar
+            0xA0..=0xD7FF
+            | 0xF900..=0xFDCF
+            | 0xFDF0..=0xFFEF
+            | 0x10000..=0x1FFFD
+            | 0x20000..=0x2FFFD
+            | 0x30000..=0x3FFFD
+            | 0x40000..=0x4FFFD
+            | 0x50000..=0x5FFFD
+            | 0x60000..=0x6FFFD
+            | 0x70000..=0x7FFFD
+            | 0x80000..=0x8FFFD
+            | 0x90000..=0x9FFFD
+            | 0xA0000..=0xAFFFD
+            | 0xB0000..=0xBFFFD
+            | 0xC0000..=0xCFFFD
+            | 0xD0000..=0xDFFFD
+            | 0xE1000..=0xEFFFD => false,
+            // iprivate
+            0xE000..=0xF8FF | 0xF0000..=0xFFFFD | 0x100000..=0x10FFFD => {
+                self.0 != IriSegmentKind::Query
+            }
+            _ => true,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn iri_encode_cyrillic() {
+        let encoder = IriReserved(IriSegmentKind::Segment);
+        let pct_string = PctString::encode("традиционное польское блюдо".chars(), encoder);
```

Comment:
I don't speak Russian, I hope that's not offensive in any ways :wink:

Reply:
I just copy and pasted from the article about pirogi in Russian, a "traditional Polish dish". Why? I was eating one at the time, lol

Reply:
I like that, it adds some character to the crate :smile: